_posts/2017-03-28-table-sql-api-update.md

```diff
@@ -0,0 +1,195 @@
+
+    42|Bob Smith|2016-07-23 16:10:11|color=12,length=200,size=200
+
+The following example illustrates how to read a CSV file and perform some data cleansing before converting it to a regular DataStream program.
+
+```scala
+// set up execution environment
+val env = StreamExecutionEnvironment.getExecutionEnvironment
+val tEnv = TableEnvironment.getTableEnvironment(env)
+
+// configure table source
+val customerSource = CsvTableSource.builder()
+  .path("/path/to/customer_data.csv")
+  .ignoreFirstLine()
+  .fieldDelimiter("|")
+  .field("id", Types.LONG)
+  .field("name", Types.STRING)
+  .field("last_update", Types.TIMESTAMP)
+  .field("prefs", Types.STRING)
+  .build()
+
+// name your table source
+tEnv.registerTableSource("customers", customerSource)
+
+// define your table program
+val table = tEnv
+  .scan("customers")
+  .filter('name.isNotNull && 'last_update > "2016-01-01 00:00:00".toTimestamp)
+  .select('id, 'name.lowerCase(), 'prefs)
+
+// convert it to a data stream
+val ds = table.toDataStream[Row]
+
+ds.print()
+env.execute()
+```
+
+The Table API comes with a large set of built-in functions that make it easy to specify  business logic using a language integrated query (LINQ) syntax. In the example above, we filter out customers with invalid names and only select those that updated their preferences recently. We convert names to lowercase for normalization. For debugging purposes, we convert the table into a DataStream and print it.
+
+The `CsvTableSource` supports both batch and stream environments. If the programmer wants to execute the program above in a batch application, all he or she has to do is to replace the environment via `ExecutionEnvironment` and change the output conversion from `DataStream` to `DataSet`. The Table API program itself doesn’t change.
+
+In the example, we converted the table program to a data stream of `Row` objects. However, we are not limited to row data types. The Table API supports all types from the underlying APIs such as Java and Scala Tuples, Case Classes, POJOs, or generic types that are serialized using Kryo. Let’s assume that we want to have regular object (POJO) with the following format instead of generic rows:
+
+```scala
+class Customer {
+  var id: Int = _
+  var name: String = _
+  var update: Long = _
+  var prefs: java.util.Properties = _
+}
+```
+We can use the following table program to convert the CSV file into Customer objects. Flink takes care of creating objects and mapping fields for us.
+
+```scala
+val ds = tEnv
+  .scan("customers")
+  .select('id, 'name, 'last_update as 'update, parseProperties('prefs) as 'prefs)
+  .toDataStream[Customer]
+```
+
+Even though Flink’s Table API is shipped with a large set of built-in functions, it is often necessary to define custom user-defined scalar functions in this example to parse the preferences into a Java Properties object. The following code snippet shows how easily we can implement a scalar function.
```

Comment:
"is often necessary to define custom user-defined scalar functions. In the above example we use a user-defined function `parseProperties`..."